src/bin/elasticsearch_exporter.rs

```diff
@@ -0,0 +1,359 @@
+#![feature(str_split_once)]
+
+#[macro_use]
+extern crate lazy_static;
+#[macro_use]
+extern crate prometheus;
+#[macro_use]
+extern crate log;
+
+use clap::Clap;
+use hyper::server::conn::AddrStream;
+use hyper::{
+    header::CONTENT_TYPE,
+    service::{make_service_fn, service_fn},
+    Body, Request, Response, Server, StatusCode,
+};
+use prometheus::{Encoder, HistogramVec, TextEncoder, TEXT_FORMAT};
+use std::convert::Infallible;
+use std::env;
+use std::error::Error as StdError;
+use std::fmt;
+use std::net::SocketAddr;
+use std::panic;
+use std::str::FromStr;
+use std::time::Duration;
+use tokio::signal;
+use tokio::sync::oneshot::{self, Receiver, Sender};
+use url::Url;
+
+use elasticsearch_exporter::{
+    CollectionLabels, Exporter, ExporterOptions, ExporterPollIntervals, Labels,
+};
+
+lazy_static! {
+    static ref HTTP_REQ_HISTOGRAM: HistogramVec = register_histogram_vec!(
+        "http_request_duration_seconds",
+        "The HTTP request latencies in seconds.",
+        &["handler"]
+    )
+    .expect("valid histogram vec metric");
+}
+
+fn build_response(status: StatusCode, body: Body) -> Response<Body> {
+    Response::builder()
+        .status(status)
+        .header(CONTENT_TYPE, TEXT_FORMAT)
+        .body(body)
+        .expect("valid Response built")
+}
+
+async fn serve_req(req: Request<Body>, options: String) -> Result<Response<Body>, Infallible> {
+    let path = req.uri().path();
+
+    let timer = HTTP_REQ_HISTOGRAM.with_label_values(&[path]).start_timer();
+
+    let response = match path {
+        "/health" | "/healthy" | "/healthz" => build_response(StatusCode::OK, Body::from("Ok")),
+        "/" => build_response(StatusCode::OK, Body::from(options.to_string())),
+
+        "/metrics" => {
+            let encoder = TextEncoder::new();
+
+            let mut buffer = vec![];
+            match encoder.encode(&prometheus::gather(), &mut buffer) {
+                Ok(_) => build_response(StatusCode::OK, Body::from(buffer)),
+                Err(e) => {
+                    error!("prometheus encoder err {}", e);
+
+                    build_response(StatusCode::INTERNAL_SERVER_ERROR, Body::empty())
+                }
+            }
+        }
+        _ => build_response(
+            StatusCode::NOT_FOUND,
+            Body::from(format!("Path {} not found", path)),
+        ),
+    };
+
+    timer.observe_duration();
+
+    Ok(response)
+}
+
+#[derive(Clap, Clone, Debug)]
+struct Opts {
```

Comment:
I'd move this into a separate file. I believe `elasticsearch_exporter.rs ` should contain only startup logic.